crates/sui-graphql-rpc/src/types/transaction_block.rs

```diff
@@ -182,6 +188,188 @@ impl TransactionBlock {
+                    query = query.filter(tx::dsl::transaction_kind.eq(*k as i16))
+                }
+
+                if let Some(c) = &filter.after_checkpoint {
+                    query = query.filter(tx::dsl::checkpoint_sequence_number.gt(*c as i64));
+                }
+
+                if let Some(c) = &filter.at_checkpoint {
+                    query = query.filter(tx::dsl::checkpoint_sequence_number.eq(*c as i64));
+                }
+
+                if let Some(c) = &filter.before_checkpoint {
+                    query = query.filter(tx::dsl::checkpoint_sequence_number.lt(*c as i64));
+                }
+
+                if let Some(a) = &filter.sign_address {
+                    let sub_query = tx_senders::dsl::tx_senders
+                        .select(tx_senders::dsl::tx_sequence_number)
+                        .filter(tx_senders::dsl::sender.eq(a.into_vec()));
+                    query = query.filter(tx::dsl::tx_sequence_number.eq_any(sub_query));
+                }
+
+                if let Some(a) = &filter.recv_address {
+                    let sub_query = tx_recipients::dsl::tx_recipients
+                        .select(tx_recipients::dsl::tx_sequence_number)
+                        .filter(tx_recipients::dsl::recipient.eq(a.into_vec()));
+                    query = query.filter(tx::dsl::tx_sequence_number.eq_any(sub_query));
+                }
+
+                if let Some(o) = &filter.input_object {
+                    let sub_query = tx_input_objects::dsl::tx_input_objects
+                        .select(tx_input_objects::dsl::tx_sequence_number)
+                        .filter(tx_input_objects::dsl::object_id.eq(o.into_vec()));
+                    query = query.filter(tx::dsl::tx_sequence_number.eq_any(sub_query));
+                }
+
+                if let Some(o) = &filter.changed_object {
+                    let sub_query = tx_changed_objects::dsl::tx_changed_objects
+                        .select(tx_changed_objects::dsl::tx_sequence_number)
+                        .filter(tx_changed_objects::dsl::object_id.eq(o.into_vec()));
+                    query = query.filter(tx::dsl::tx_sequence_number.eq_any(sub_query));
+                }
+
+                if let Some(txs) = &filter.transaction_ids {
+                    let digests: Vec<_> = txs.iter().map(|d| d.to_vec()).collect();
+                    query = query.filter(tx::dsl::transaction_digest.eq_any(digests));
+                }
+
+                query
+            })
+            .await?;
+
+        let mut conn = Connection::new(prev, next);
+
+        for stored in results {
+            let cursor = Cursor::new(stored.cursor()).encode_cursor();
+            let transaction = TransactionBlock::try_from(stored)?;
+            conn.edges.push(Edge::new(cursor, transaction));
+        }
+
+        Ok(conn)
+    }
+}
+
+impl TransactionBlockFilter {
+    pub(crate) fn merge(self, other: Self) -> Option<Self> {
```

Comment:
So is this trying to merge the ids from different kinds of filters?

Reply:
It produces a filter that corresponds to the intersection of `self` and `other`, as long as that's possible.  It's not possible if the resulting filter is inconsistent in some static way (needs to match two distinct signer or receiver addresses, etc, etc).

We have to fail in those inconsistent cases because the representation of the filter isn't "closed under intersection" (there are some intersections that can't be represented by the type), but we take advantage of that fact to avoid a database round-trip (if `merge` returns `None` then we return an empty result).

I'll add a doc comment to this function.